lib/src/persistent_queue.dart

```diff
@@ -0,0 +1,133 @@
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+import 'package:matomo_tracker/src/dispatch_settings.dart';
+import 'package:matomo_tracker/src/local_storage/local_storage.dart';
+
+class PersistentQueue extends DelegatingQueue<Map<String, String>> {
+  PersistentQueue._(this._storage, Queue<Map<String, String>> base)
+      : _needsSave = false,
+        super(base);
+
+  static Future<PersistentQueue> load({
+    required LocalStorage storage,
+    required PersistenceFilter onLoadFilter,
+  }) async {
+    final actionsData = await storage.loadActions();
+    final actions = decode(actionsData)..retainWhere(onLoadFilter);
+    final queue = PersistentQueue._(storage, Queue.of(actions));
+    await queue.save();
+    return queue;
+  }
+
+  @visibleForTesting
+  static List<Map<String, String>> decode(String? actionsData) {
+    if (actionsData != null) {
+      return (json.decode(actionsData) as List)
+          .cast<Map<String, dynamic>>()
+          .map<Map<String, String>>((element) => element.cast())
+          .toList();
+    } else {
+      return [];
+    }
+  }
+
+  final LocalStorage _storage;
+  bool _needsSave;
+  Completer<void>? _saveInProgress;
+  @visibleForTesting
+  bool get saveInProgress => _saveInProgress != null;
```

Comment:
Wouldn't the `Lock` preferable to this?

Reply:
Hm I don' think so. The `Lock` is used to execute the `func`s in order and make subsequent calls wait for the last `func` to finish before starting to execute the next. In this case here, there are no subsequent calls but rather just one future that after `await _storage.storeActions(data)`, checks if something changed while `awaiting` and if so, makes another `storeActions` attempt. In other words: we don't care if there are *multiple* changes while `awaiting` and just need to know if there was at least *one* change and then store the most current state.